cmd/util/ledger/migrations/ordered_map_migration.go

```diff
@@ -0,0 +1,272 @@
+	panic("unreachable")
+}
+
+func (RawStorable) NeedsStoreTo(_ atree.Address) bool {
+	panic("unreachable")
+}
+
+func (RawStorable) IsResourceKinded(_ *interpreter.Interpreter) bool {
+	panic("unreachable")
+}
+
+func (v RawStorable) Transfer(
+	interpreter *interpreter.Interpreter,
+	_ func() interpreter.LocationRange,
+	_ atree.Address,
+	remove bool,
+	storable atree.Storable,
+) interpreter.Value {
+	panic("unreachable")
+}
+
+func (v RawStorable) Clone(_ *interpreter.Interpreter) interpreter.Value {
+	panic("unreachable")
+}
+
+func (RawStorable) DeepRemove(_ *interpreter.Interpreter) {
+	// NO-OP
+}
+
+func (r RawStorable) Encode(enc *atree.Encoder) error {
+	return enc.CBOR.EncodeRawBytes(r)
+}
+
+func (r RawStorable) ByteSize() uint32 {
+	return uint32(len(r))
+}
+
+func (r RawStorable) StoredValue(storage atree.SlabStorage) (atree.Value, error) {
+	return r, nil
+}
+
+func (r RawStorable) ChildStorables() []atree.Storable {
+	return nil
+}
+
+func (r RawStorable) Storable(_ atree.SlabStorage, _ atree.Address, _ uint64) (atree.Storable, error) {
+	return r, nil
+}
+
+func splitPayloads(inp []ledger.Payload) (fvmPayloads []ledger.Payload, storagePayloads []ledger.Payload, slabPayloads []ledger.Payload) {
+	for _, p := range inp {
+		if state.IsFVMStateKey(
+			string(p.Key.KeyParts[0].Value),
+			string(p.Key.KeyParts[1].Value),
+			string(p.Key.KeyParts[2].Value),
+		) {
+			fvmPayloads = append(fvmPayloads, p)
+			continue
+		}
+		if bytes.HasPrefix(p.Key.KeyParts[2].Value, []byte(atree.LedgerBaseStorageSlabPrefix)) {
+			slabPayloads = append(slabPayloads, p)
+			continue
+		}
+		// otherwise this is a storage payload
+		storagePayloads = append(storagePayloads, p)
+	}
+	return
+}
+
+func (m *OrderedMapMigration) initialize(payload []ledger.Payload) ([]ledger.Payload, error) {
+	fvmPayloads, storagePayloads, slabPayloads := splitPayloads(payload)
+
+	m.ledgerView = NewView(append(fvmPayloads, slabPayloads...))
+	m.initPersistentSlabStorage(m.ledgerView)
+	m.initIntepreter()
+	return storagePayloads, nil
+}
+
+func (m *OrderedMapMigration) migrate(storagePayloads []ledger.Payload) ([]ledger.Payload, error) {
+	groupedByOwnerAndDomain := make(map[string](map[string][]Pair))
+
+	for _, p := range storagePayloads {
+		owner, entry :=
+			string(p.Key.KeyParts[0].Value),
+			string(p.Key.KeyParts[2].Value)
+		// if the entry doesn't contain a separator, we just ignore it
+		// since it is storage metadata and does not need migration
+		if !strings.Contains(entry, "storage\x1f") &&
+			!strings.Contains(entry, "public\x1f") &&
+			!strings.Contains(entry, "contract\x1f") &&
+			!strings.Contains(entry, "private\x1f") {
+			m.Log.Info().Msgf("Ignoring key in storage payloads: %s", entry)
+			continue
+		}
+		splitEntry := strings.Split(entry, "\x1f")
+		domain, key := splitEntry[0], splitEntry[1]
+		value := p.Value
+
+		domainMap, domainOk := groupedByOwnerAndDomain[owner]
+		if !domainOk {
+			domainMap = make(map[string][]Pair)
+		}
```

Comment:
This assumes if we don't have any value on a specific domain for an account we won't create an empty map for it, just checking this won't cause any issue adding a new value later to that domain? or cadence would expect the map exists for all domains for all accounts

Reply:
cc @turbolent this is a good question, if an account has nothing in `storage` for example, is that represented by an empty map, or should there be nothing associated with the `"storage"` key? If something is later added to that account's `storage`, is the map created when its first element is added?